Parser/pegen/pegen.c

```diff
@@ -756,24 +756,18 @@ _PyPegen_expect_token(Parser *p, int type)
 expr_ty
 _PyPegen_expect_soft_keyword(Parser *p, const char *keyword)
 {
-    if (p->mark == p->fill) {
-        if (_PyPegen_fill_token(p) < 0) {
-            p->error_indicator = 1;
-            return NULL;
-        }
-    }
-    Token *t = p->tokens[p->mark];
-    if (t->type != NAME) {
+    expr_ty res = _PyPegen_name_token(p);
+    if (!res) {
         return NULL;
     }
-    char* s = PyBytes_AsString(t->bytes);
+    const char *s = PyUnicode_AsUTF8(res->v.Name.id);
     if (!s) {
+        p->error_indicator = 1;
         return NULL;
     }
     if (strcmp(s, keyword) != 0) {
         return NULL;
```

Comment:
I still think you need to reset p->mark here, otherwise the parser becomes too hard to reason about.

Basically, I think you should just revert this part of the diff; I'm also not happy with the potential work done by PyUnicode_AsUTF8() if there's a non-ASCII letter in the identifier.

Reply:
You know best! Reverting.